File: Algorithm/leetcode300.cpp
/* Leetcode 300 Longest Increasing Subsequence */
/* Given an unsorted array of integers, find the length of longest increasing subsequence. */
//Note:
//	There may be more than one LIS combination, it is only necessary for you to return the length.
//	Your algorithm should run in O(n2) complexity.
//Follow up: Could you improve it to O(n log n) time complexity?

//解题思路：动态规划
//子问题设置：以ak(k=0,1,...,n-1)为终点的上升子序列最大长度

#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
	int lengthOfLIS(vector<int>& nums) {
		int n = nums.size();
		if (n == 0) { return 0; }
		int maxLength = 1;
		vector<int> dp(n, 1);					//dp[i]表示以nums[i]为终点的上升子序列最大长度
		for (int i = 1; i < n; i++) {
			for (int j = 0; j < i; j++) {
				if (nums[i] > nums[j]) {
					dp[i] = max(dp[i], dp[j] + 1);
					maxLength = max(dp[i], maxLength);
				}
			}
		}
		return maxLength;
	}
};		//时间复杂度O(n^2)

int main() {
	Solution s;
	vector<int> v;
	v.push_back(1); v.push_back(3); v.push_back(6);
	v.push_back(7); v.push_back(9); v.push_back(4);
	v.push_back(10); v.push_back(5); v.push_back(6);
	s.lengthOfLIS(v);
	return 0;
}

File: Algorithm/Leetcode300.java
import java.util.Arrays;

public class Leetcode300 {
    /* Leetcode 300 Longest Increasing Subsequence */

    /**
     * Given an unsorted array of integers, find the length of longest increasing subsequence.
     * Note:
     * There may be more than one LIS combination, it is only necessary for you to return the length.
     * Your algorithm should run in O(n2) complexity.
     */
    public static void main(String[] args) {
        //这题之前应该确实是用C++做过，然后还记得大致的方向应该是DP
        //不过具体的就不怎么记得了
        //然后这次又刚好在Leetcode的每日一题上刷到了，而且刚好又很久没写JAVA了感觉（前几天都在写前端）
        // 所以我觉得确实可以再做一下
        Solution solution = new Solution();
        int[] arr = {10, 9, 2, 5, 3, 7, 101, 18};
        solution.lengthOfLIS(arr);
    }
}

class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];          //dp[i]表示以第i个字符为结尾的LIS
        int ret = 0;
        Arrays.fill(dp, 1);
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    if (dp[j] + 1 > dp[i]) {
                        dp[i] = dp[j] + 1;
                    }
                }
            }
        }
        for (int i = 0; i < n; i++) {       //dp[n-1]不一定是最大值
            if (dp[i] > ret) {
                ret = dp[i];
            }
        }
        return ret;
    }
}